FAL-3449 - Taxonomy view management apis [WIP]

rpenido opened the pull request.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/serializers.py:
```diff
@@ -0,0 +1,19 @@
+from rest_framework import serializers
+
+
+class TaxonomySerializer(serializers.Serializer):
```
Using [ModelSerializer](https://www.django-rest-framework.org/api-guide/serializers/#modelserializer)s will reduce the lines of code required here.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/serializers.py:
```diff
@@ -0,0 +1,19 @@
+    allow_free_text = serializers.BooleanField()
+    system_defined = serializers.BooleanField()
+    visible_to_authors = serializers.BooleanField()
+    _object_tag_class = serializers.CharField()
```
No need to serialize this field -- it's for backend use only, and so isn't needed by users of the REST API.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,35 @@
+"""
+from rest_framework.viewsets import ModelViewSet
+
+from openedx.core.lib.api.view_utils import DeveloperErrorViewMixin
```
By design, we can't depend on code in edx-platform in this repo -- it needs to stand alone. cf https://github.com/openedx/openedx-learning/blob/main/docs/decisions/0001-purpose-of-this-repo.rst

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/urls.py:
```diff
@@ -0,0 +1,14 @@
+
+urlpatterns = [
+    path('', include(router.urls))
+]
```
Yep, this is correct, but we'll also need a couple more:

* a `openedx_tagging/core/tagging/rest_api/urls.py` file [like this one](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/demographics/rest_api/urls.py) that [includes](https://docs.djangoproject.com/en/4.0/ref/urls/#include) these v1 URLs.
* a `openedx_tagging/core/tagging/urls.py` file that [includes](https://docs.djangoproject.com/en/4.0/ref/urls/#include) the rest_api urls, gives them a root path and sets their [url namespace](https://docs.djangoproject.com/en/4.0/topics/http/urls/#url-namespaces) to `oel_tagging`.

rpenido commented on projects/urls.py:
```diff
@@ -5,9 +5,10 @@
 urlpatterns = [
-    path("admin/doc/", include("django.contrib.admindocs.urls")),
-    path("admin/", admin.site.urls),
+    # path("admin/doc/", include("django.contrib.admindocs.urls")), # Check error
+    # path("admin/", admin.site.urls), # Check error
```
With the lines above I got a `LookupError: No installed app with label 'admin'.` and can't run the tests.

pomegranited replied:
The issue is with the [`test_settings.py` file](https://github.com/open-craft/openedx-learning/blob/main/test_settings.py#L36-L38) -- uncomment those lines to include `'django.contrib.admin'` and `'django.contrib.admindocs'` in the test `INSTALLED_APPS.

pomegranited commented on openedx_tagging/core/tagging/rules.py:
```diff
@@ -28,7 +28,7 @@ def can_change_taxonomy(user: User, taxonomy: Taxonomy = None) -> bool:
     """
     return is_taxonomy_admin(user) and (
-        not taxonomy or not taxonomy or (taxonomy and not taxonomy.system_defined)
+        not taxonomy or (taxonomy and not taxonomy.system_defined)
```
Oh wow, thanks for spotting that!

pomegranited commented on projects/urls.py:
```diff
@@ -5,9 +5,10 @@
     path("media_server/", include("openedx_learning.contrib.media_server.urls")),
     path("rest_api/", include("openedx_learning.rest_api.urls")),
-    path('__debug__/', include('debug_toolbar.urls')),
+    path("rest_api2/", include("openedx_tagging.core.tagging.urls")),
```
Let's put this under a different URL, and add a namespace:
```suggestion
    path("tagging/rest_api/", include("openedx_tagging.core.tagging.urls", namespace="oel_tagging")),
```

This change (plus the [added `app_name` change suggested above](https://github.com/open-craft/openedx-learning/pull/4/files#diff-c1ed6a4e1641edb5b1f7ad4f61be866a45c8784463eff6b7d7f54adc8775d891R8)) mean you need to prefix the URL names requested by `reverse()` with this namespace:

```
url = reverse("oel_tagging:taxonomy-list")
```

pomegranited commented on openedx_tagging/core/tagging/urls.py:
```diff
@@ -0,0 +1,11 @@
+from django.urls import path, include
+
+from .rest_api import urls
+
```
We need an `app_name` here in order to use [the namespace I recommended below](https://github.com/open-craft/openedx-learning/pull/4/files#diff-736f67c058debd0bdd57a6bf62f4c9caf366ca6223e987fd3996a989b0d2fe7dR12):

```suggestion

app_name = "oel_tagging"
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,266 @@
+
+
+def _has_perm_view_taxonomy(name, *args, **kwargs):
+    return name == "oel_tagging.view_taxonomy"
```
You shouldn't patch permissions in these tests.. if you find you have to, then that means there's either something awry with the names I've used in my `rules.py` (like, maybe I didn't name them properly for DjangoModelPermission?), or the `permission_classes` configured on your views aren't working.

But if you use the `staff` user created below, then don't your tests work without patching?

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,266 @@
+            response = self.client.get(url)
+            self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_detail_taxonomy(self):
```
Copy-paste error: Because this method is named `test_detail_taxonomy`, the same name as the test above, only the second definition of this method will be run.

I tried to figure out how to enable test coverage on the `tests/*` files too -- thought I just needed to add `tests` to the list of `source` dirs in [`.coveragerc`](https://github.com/open-craft/openedx-learning/blob/rpenido/fal-3449-taxonomy-view-management-apis/.coveragerc#L4-L6), but that didn't work for some reason. Maybe you can figure out why not?

rpenido replied:
The `.coveragerc` is being overrided by [`tox.ini`](https://github.com/open-craft/openedx-learning/blob/7aa5b005e7c8d872c9d1bed6b9fe54289504ef95/tox.ini#L34) config. Now the coverage are looking on `tests/*` too.

Should we remove `.coveragerc`?

pomegranited replied:
Good diagnosis, but no, not yet -- I'd like to keep `.coveragerc` in place so I can run `make quality` locally.
(I can't run `tox` locally because of the mysql testing stuff, which is critical for us before merging, but will take some work to update.)

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,266 @@
+    #     ):
+    #         self.client.force_authenticate(user=self.user)
+    #         response = self.client.put(url, {"name": "new name"})
+    #         self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
```
Yes, please include this test.

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,266 @@
+            taxonomy = Taxonomy.objects.get(id=taxonomy.id)
+            self.assertEqual(taxonomy.name, "new name")
+
+    def test_delete_taxonomy_permission_anon(self):
+        url = reverse("taxonomy-detail", kwargs={"id": self.taxonomy.id})
+
+        response = self.client.delete(url)
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_delete_taxonomy_permission_user(self):
+        url = reverse("taxonomy-detail", kwargs={"id": self.taxonomy.id})
+
+        self.client.force_authenticate(user=self.user)
+        response = self.client.delete(url)
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_delete_taxonomy(self):
+        # Happy path
+
+        taxonomy = Taxonomy.objects.create(name="test delete taxonomy")
+        taxonomy.save()
+
+        url = reverse("taxonomy-detail", kwargs={"id": taxonomy.id})
+        with patch.object(
+            self.user, "has_perm", _has_perm_delete_taxonomy
+        ):
+            self.client.force_authenticate(user=self.user)
+            response = self.client.delete(url)
+            self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+            with self.assertRaises(Taxonomy.DoesNotExist):
+                Taxonomy.objects.get(id=taxonomy.id)
```
Please consolidate your tests wherever possible using ddt, 

e.g.

```suggestion
    @ddt.data(
        (None, status.HTTP_403_FORBIDDEN),
        ('user', status.HTTP_403_FORBIDDEN),
        ('staff', status.HTTP_204_NO_CONTENT),
    )
    @ddt.unpack
    def test_delete_taxonomy(self, user_attr, expected_status):
        url = reverse("oel_tagging:taxonomy-detail", kwargs={"id": self.taxonomy.id})

        user = getattr(self, user_attr) if user_attr else None
        if user:
            self.client.force_authenticate(user=user)
        response = self.client.delete(url)
        self.assertEqual(response.status_code, expected_status)

        # If we were able to delete the taxonomy, check that it's really gone
        if status.is_success(expected_status):
            response = self.client.get(url)
            self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```

pomegranited replied:
Note that the last line of my suggested change fails, but it shouldn't -- the `detail` view should return a 404 when the taxonomy doesn't exist.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,69 @@
+
+    serializer_class = TaxonomySerializer
+    lookup_field = "id"
+    permission_classes = [DjangoModelPermissions]
```
If you don't need any custom permissions, then you don't need a `permissions.py` file either.

But if you do end up needing custom permissions, I recommend the approach [described here](https://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions):

> To use custom model permissions, override `DjangoModelPermissions` and set the `.perms_map` property. Refer to the source code for details.

This post describes it well: https://nezhar.com/blog/django-rest-framework-permissions-in-depth/

rpenido commented on openedx_tagging/core/tagging/rest_api/v1/serializers.py:
```diff
@@ -0,0 +1,72 @@
+    enabled = serializers.BooleanField(required=False)
+
+class ClassField(serializers.Field):
+    """
```
@pomegranited Do you think that we should create a now Field type in the model to handle this logic?

pomegranited replied:
Sorry, I should have said: This `taxonomy_class` field doesn't need to be serialized -- it's for backend use only, so the frontend doesn't need to see or update it.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/permissions.py:
```diff
@@ -0,0 +1,30 @@
+class TaxonomyObjectPermissions(DjangoObjectPermissions):
+    perms_map = {
+        'GET': ['%(app_label)s.view_%(model_name)s'],
+        # 'GET': [],
```
nit:
```suggestion
```

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/serializers.py:
```diff
@@ -0,0 +1,72 @@
+            "allow_free_text",
+            "system_defined",
+            "visible_to_authors",
+            "taxonomy_class",
```
Don't need to serialize this field:
```suggestion
```

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/serializers.py:
```diff
@@ -0,0 +1,72 @@
+        return value
+
+class TaxonomySerializer(serializers.ModelSerializer):
+    taxonomy_class = ClassField(required=False, allow_null=True)
```
```suggestion
```

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/serializers.py:
```diff
@@ -0,0 +1,72 @@
+        raise NotImplementedError(
+            "Cannot save a taxonomy through serializer. Use the create_taxonomy or"
+            "update_taxonomy function instead."
+        )
```
No need to disable these.. see my comments [below on the ModelViewSet](https://github.com/open-craft/openedx-learning/pull/4/files/a0fedd36bbe06ef0a4c00466c8c9570f3c35616d#diff-215d20c663dcd5727c14740a72d09fe6fd22e3a92fd08db46036e2e799699917R174).
```suggestion
```

pomegranited commented on openedx_tagging/core/tagging/api.py:
```diff
@@ -68,6 +68,21 @@ def get_taxonomies(enabled=True) -> QuerySet:
+    taxonomy = Taxonomy.objects.get(id=id)
+    taxonomy.delete()
+
+
```
I don't think we need these.. it's ok to just let the ModelViewSet modify/delete the Taxonomy models.
```suggestion
```

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,180 @@
+        """
+        Delete a taxonomy.
+        """
+        delete_taxonomy(instance.id)
```
I don't think these methods should be re-implemented. We should use `api.create_taxonomy` when creating, but it's ok to let the default ModelViewSet handle updates and destroy, no need to create or use API methods for update/delete.

Also: it's better to override just `perform_create` instead of the whole `create` method.

```suggestion
    def perform_create(self, serializer):
        """
        Create a new taxonomy.
        """
        serializer.instance = create_taxonomy(**serializer.validated_data)
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_serializer.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@ddt.ddt
+class TestTaxonomySerializer(TestCase):
```
Ok to remove this test -- I've recommended removing the method overrides, and we don't need to serialize taxonomy_class, and so the serializers are tested by your view tests.

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+            elif enabled in [True, "True", 1, "1"]:
+                self.assertEqual(len(response.data), 2)
+            else:
+                self.assertEqual(len(response.data), 1)
```
Could you put the number of taxonomies expected by the result in the ddt.data instead of listing them each here?
```suggestion
        # If we were able to list the taxonomies, check that we got the expected number back
        if status.is_success(expected_status):
            assert len(response.data) == expected_count
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+            response = self.client.get(url, {"enabled": enabled})
+        else:
+            response = self.client.get(url)
+        self.assertEqual(response.status_code, expected_status)
```
style nit for all tests: I've been using simpler asserts:
```suggestion
        assert response.status_code == expected_status
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+
+        user = getattr(self, user_attr) if user_attr else None
+        if user:
+            self.client.force_authenticate(user=user)
```
nit:

```suggestion
        if user_attr:
            user = getattr(self, user_attr)
            self.client.force_authenticate(user=user)
```

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/permissions.py:
```diff
@@ -0,0 +1,30 @@
+        'DELETE': ['%(app_label)s.delete_%(model_name)s'],
+    }
+
+    def has_permission(self, request, view):
+        # Workaround to allow 'retrieve' view to pass through the
+        # method GET permission check. Actual object permission check
+        # is done in the get_object() method.
+        if view.action == 'retrieve':
+            return bool(request.user and request.user.is_authenticated)
+        else:
+            return super().has_permission(request, view)
+
+
```
I don't understand why this is needed.. it looks like the [DjangoObjectPermissions](https://www.django-rest-framework.org/api-guide/permissions/#djangoobjectpermissions) use [DjangoModelPermissions](https://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions) when there's no object in the query, right?

And normal users cannot view taxonomies, only taxonomy admins:

```python
request.user.has_perm("oel_tagging.view_taxonomy")
```

So I think this whole block can go?

```suggestion
```

rpenido replied:
From `rules.py`: https://github.com/open-craft/openedx-learning/blob/a0fedd36bbe06ef0a4c00466c8c9570f3c35616d/openedx_tagging/core/tagging/rules.py#L16-L22

Can you please confirm if normal users cannot view taxonomies? If so, we need to update the rule.


> don't understand why this is needed.. it looks like the [DjangoObjectPermissions](https://www.django-rest-framework.org/api-guide/permissions/#djangoobjectpermissions) use [DjangoModelPermissions](https://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions) when there's no object in the query, right?

Yes!
When we do a call to ` GET taxonomy/:id`, first it calls `has_permission` (without object, to validate the GET request permission) and only after that calls `has_object_permission`. The problem was that a normal user doesn't have GET general permission, so I needed this workaround.

This code can go if a normal user cannot retrieve a taxonomy.

pomegranited replied:
Yes, this code can go -- it's just taxonomy administrators who can access taxonomies.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/permissions.py:
```diff
@@ -0,0 +1,30 @@
+        # 'GET': [],
+        'OPTIONS': [],
+        # 'HEAD': ['%(app_label)s.view_%(model_name)s'],
+        'HEAD': [],
```
Should be:
```suggestion
        'HEAD': ['%(app_label)s.view_%(model_name)s'],
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+
+    @ddt.data(
+        (None, {"enabled": True}, status.HTTP_403_FORBIDDEN),
+        ("user", {"enabled": True}, status.HTTP_200_OK),
+        ("user", {"enabled": False}, status.HTTP_404_NOT_FOUND),
```
Nope.. normal users aren't allowed to view taxonomies, only taxonomy admins.

When you remove `has_permission` as [suggested above](https://github.com/open-craft/openedx-learning/pull/4/files/a0fedd36bbe06ef0a4c00466c8c9570f3c35616d#diff-091bcd6aa37fc62100ec5e982af82be0886e0ea9025f8bc15bea40bd67f978adR21), this is reflected correctly, and should be:

```suggestion
        ("user", {"enabled": True}, status.HTTP_403_FORBIDDEN),
        ("user", {"enabled": False}, status.HTTP_403_FORBIDDEN),
```

rpenido replied:
Commented above

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+        ("user", {"enabled": True}, status.HTTP_200_OK),
+        ("user", {"enabled": False}, status.HTTP_404_NOT_FOUND),
+        ("staff", {"enabled": True}, status.HTTP_200_OK),
+        ("staff", {"enabled": True}, status.HTTP_200_OK),
```
missed one:
```suggestion
        ("staff", {"enabled": False}, status.HTTP_200_OK),
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+                "oel_tagging:taxonomy-detail", kwargs={"id": response.data["id"]}
+            )
+            response = self.client.get(url)
+            self.assertGreaterEqual(response.data.items(), create_data.items())
```
Ditto

```suggestion
            self.check_taxonomy(response.data, **create_data)
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+        self.assertEqual(response.status_code, expected_status)
+
+    @ddt.data(
+        (None, {"enabled": True}, status.HTTP_403_FORBIDDEN),
```
For completeness, you can add:
```suggestion
        (None, {"enabled": True}, status.HTTP_403_FORBIDDEN),
        (None, {"enabled": False}, status.HTTP_403_FORBIDDEN),
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, expected_status)
+
+        if status.is_success(expected_status):
+            response = self.client.get(url)
+            self.assertGreaterEqual(response.data.items(), create_data.items())
```
I'm confused.. why do we need to call `get` again?
And instead of just checking that the data fields are the right length, could you check the returned data itself? We can add a helper method for this. (Applies everywhere there's an `assertGreaterEqual`)

```suggestion
        if status.is_success(expected_status):
            self.check_taxonomy(response.data, taxonomy.id, **create_data)
            
    def check_taxonomy(self, data, id, name="", description=None, enabled=True, required=False, allow_multiple=False, allow_free_text=False):
        """
        Check that the data dict contains the expected fields.
        """
        assert data['id'] == id
        assert data['name'] == name
        assert data['enabled'] == enabled
        assert data['description'] == description
        assert data['required'] == required
        assert data['allow_multiple'] == allow_multiple
        assert data['allow_free_text'] == allow_free_text
        assert data['system_defined'] == False
```

rpenido replied:
My bad here! We don't need to call get twice!

But the `assertGreaterEqual` in that way doesn't compare the count. It compares if the second dict is a subset of the first (the data itself).

```python
a = {
  "foo": "foo",
  "bar": "bar",
}
b = {
  "foo": "foo",
}
c = {
  "foo": "fzz",
}

self.assertGreaterEqual(a.items(), b.items()) # This pass
self.assertGreaterEqual(a.items(), c.items()) # This FAILS
# The order matters
self.assertGreaterEqual(b.items(), a.items()) # This FAILS
```

I'm using this because the output usually is bigger than the input (because of the optional fields). But the `check_taxonomy` with defaults also works.

Do you prefer the explicit check?

pomegranited replied:
Ah cool, I wasn't aware that this was how assertGreaterEqual worked with dicts!

But yes please, I think it's clearer to verify all the fields on the taxonomy are as expected.

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+
+        # If we were able to create the taxonomy, check if it was created
+        if status.is_success(expected_status):
+            self.assertGreaterEqual(response.data.items(), create_data.items())
```
As above:
```suggestion
          self.check_taxonomy(response.data, **create_data)
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+            "enabled": False,
+            "required": True,
+            "allow_multiple": True,
+            "taxonomy_class": 'openedx_tagging.core.tagging.models.Taxonomy'
```
Since we're not going to set taxonomy_class with the API, this test can be removed.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,180 @@
+
+    def get_queryset(self):
+        """
+        Return a list of taxonomies. If you want the disabled taxonomies, pass enabled=False. If you want all taxonomies (both enabled and disabled), pass enabled=None.
```
AFAIK, we can't really pass `None` in a query parameter.. we'd just omit `enabled` from the query.

```suggestion
        Return a list of taxonomies.
        
        Returns all taxonomies by default.
        If you want the disabled taxonomies, pass enabled=False.
        If you want the enabled taxonomies, pass enabled=True.
```

pomegranited commented on projects/urls.py:
```diff
@@ -9,5 +9,6 @@
     path("admin/", admin.site.urls),
     path("media_server/", include("openedx_learning.contrib.media_server.urls")),
     path("rest_api/", include("openedx_learning.rest_api.urls")),
+    path("tagging/rest_api/", include("openedx_tagging.core.tagging.urls", namespace="oel_tagging")),
```
Ah, I steered you wrong here.. you don't have to specify the namespace here or the app_name above. Since the tagging app already has a name, its namespace is used by default:

```suggestion
    path("tagging/rest_api/", include("openedx_tagging.core.tagging.urls")),
```

pomegranited commented on openedx_tagging/core/tagging/urls.py:
```diff
@@ -0,0 +1,10 @@
+
+from .rest_api import urls
+
+app_name = "oel_tagging"
```
Ok to remove this, see [below](https://github.com/open-craft/openedx-learning/pull/4/files/a0fedd36bbe06ef0a4c00466c8c9570f3c35616d#diff-736f67c058debd0bdd57a6bf62f4c9caf366ca6223e987fd3996a989b0d2fe7dR12):
```suggestion
```

pomegranited replied:
Bump :)

rpenido replied:
When I removed this my tests failed:
```
django.urls.exceptions.NoReverseMatch: 'oel_tagging' is not a registered namespace
```

pomegranited commented on test_settings.py:
```diff
@@ -33,8 +33,9 @@ def root(*args):
-    #    'django.contrib.admindocs',
+    'django.contrib.admin',
+    'django.contrib.admindocs',
+    'debug_toolbar',
```
If you remove this from the test INSTALLED_APPS, then can you get away with not adding `django-debug-toolbar` to the test requirements?

It's ok if these need to be there, I'm just curious.

rpenido replied:
If I comment 'django.contrib.admin' and 'django.contrib.admindocs' I get de following error in the tests:
```
FAILED tests/openedx_tagging/core/tagging/test_views.py::TestTaxonomyViewSet::test_create_taxonomy_1__None__403_ - LookupError: No installed app with label 'admin'.
```

The `django-debug-toolbar` was because I was getting an error in the following line in the CI:
https://github.com/open-craft/openedx-learning/blob/107c1f6db4c9758186b09011d513a899c14bd5e1/projects/urls.py#L13

CI Error: https://github.com/open-craft/openedx-learning/actions/runs/5599880930/jobs/10241446948?pr=4#step:7:450

The local test runs just fine.

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+        user = getattr(self, user_attr) if user_attr else None
+        if user:
+            self.client.force_authenticate(user=user)
+        response = self.client.patch(url, {"name": "new name"})
```
Ditto here to test that `system_defined` cannot be changed here.

pomegranited replied:
Bump :)

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+        user = getattr(self, user_attr) if user_attr else None
+        if user:
+            self.client.force_authenticate(user=user)
+        response = self.client.put(url, {"name": "new name"})
```
Could you also try updating `system_defined` here? It shouldn't change anything (see below):
```suggestion
        response = self.client.put(url, {"name": "new name", "system_defined": True})
```

rpenido replied:
I'm not sure if I understood you here.

Are you asking to remove the `test_update_taxonomy_system_defined` and add the proposed line here (in `test_update_taxonomy`?

pomegranited replied:
Nope, not exactly.. I wanted to verify that you cannot update the `system_taxonomy` field using the REST API.

So please add/update tests to verify:

* if you create a taxonomy via the API, it will set `system_defined=False`
* if you `put` a change to taxonomy that tries to set  `system_defined=True`, the request should succeed, but the field won't be updated -- which you can see when you re-fetch the taxonomy.
* same as ^ for [`patch`](https://github.com/open-craft/openedx-learning/pull/4#discussion_r1267530631)

rpenido replied:
Now I got it!
We already have one of these tests:

Create:

https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/tests/openedx_tagging/core/tagging/test_views.py#L161-L171

We have this update, but it is another Test Case:
https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/tests/openedx_tagging/core/tagging/test_views.py#L203-L212

(It returns 403 because of the rule here:)
https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/openedx_tagging/core/tagging/rules.py#L25-L32

--

I will create a new test for the condition that you mentioned here for UPDATE and PATCH!

pomegranited replied:
@rpenido Yay! We're on the same page now. Your `test_update_taxonomy_system_defined` test is exactly what I needed.

> (It returns 403 because of the rule here:)

👍  yep, that's correct -- the REST API should not allow updating of system-defined taxonomies, and your existing tests are sufficient for testing this.

All good here!

pomegranited commented on tests/openedx_tagging/core/tagging/test_views.py:
```diff
@@ -0,0 +1,299 @@
+            response = self.client.get(url)
+            self.assertEqual(response.data["name"], "new name")
+            self.assertEqual(response.data["enabled"], False)
+            self.assertEqual(response.data["description"], "taxonomy description")
```
Can use the `check_taxonomy` helper function we defined above.

rpenido replied:
I was trying to check if a partial update only changes one field (and does not get the defaults and override the data). 

Do you think this is a necessary check?

Using the `check_taxonomy` will compare against the default, not the previous value (unless I explicitly define the parameter). Was that your proposal?

pomegranited replied:
Yes, it's worthwhile checking that `patch` only updates what we think it does.
But to do this properly, you need to check the value of fields which you _haven't updated with the PATCH_ as well.
The `check_taxonomy` method I proposed uses the default values expected for a new taxonomy in its method kwargs, which you can override by passing in your `**update_args` or whatever you sent to the `put`/`patch`.

A simplified example:

```python
def check_taxonomy(data, id, description="", allow_free_text=False, enabled=True):
   assert data["id"] == id
   assert data["description"] == description
   assert data["enabled"] == enabled
   assert data["allow_free_text"] == allow_free_text

update_args = { "allow_free_text" = True, description = "new description"}
check_taxonomy(data, taxonomy.id, **update_args)
# ^ this will compare:
assert data["description"] ==  "new description"
assert data["enabled"] == True  # falls back to check_taxonomy param's default, because it's not in update_args
assert data["allow_free_text"] == True
```

Does that make sense?

rpenido replied:
Sure!

pomegranited commented on tox.ini:
```diff
@@ -31,7 +31,7 @@ match-dir = (?!migrations)
 [pytest]
 DJANGO_SETTINGS_MODULE = test_settings
-addopts = --cov openedx_learning --cov openedx_tagging --cov-report term-missing --cov-report xml
+addopts = --cov openedx_learning --cov openedx_tagging --cov tests --cov-report term-missing --cov-report xml
```
Brilliant! Thank you for finding this :)

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,180 @@
+
+
+    serializer_class = TaxonomySerializer
+    lookup_field = "id"
```
Ok to use the default `pk` here -- and wherever we're looking for an `"id"` in the return data, it's ok if that's `pk` too.

pomegranited reviewed (changes requested):
Hi @rpenido -- This is a good start, but I think it can be leaner.
Please see my suggestions, and feel free to push back if you disagree.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,180 @@
+
+class TaxonomyView(ModelViewSet):
+    """
+    View to list, create, retrieve, update, or delete Taxonomies.
```
This is great! No such thing as too much docs for a REST API :)

rpenido commented:
Thank you for the review, @pomegranited! The code is a lot cleaner now!

pomegranited committed:
Allow custom Taxonomy, ObjectTag subclasses to customize tagging behavior (#62)

Adds support for custom Taxonomy subclasses to be stored against a Taxonomy, to be used by the python API when instantiating and returning Taxonomies.

Also adds minimal support for ObjectTag subclasses. However, these are not stored against the ObjectTag instances; they can be instantiated by the Taxonomy subclasses if and when needed.

Related:
* docs: updates decisions to reflect this change
* feat: adds api.get_taxonomy, which returns a Taxonomy cast to its subclass, when set
* refactor: adds _check_taxonomy, _check_tag, and _check_object methods to the Taxonomy class, which can be overridden by subclasses when validating ObjectTags

Added to support system-defined Taxonomies:
* feat: adds un-editable Taxonomy.system_defined field so that system taxonomies can store this field and ensure no one edits them. 
* feat: adds Taxonomy.visible_to_authors, which is needed for fully automated tagging.

Cleanup changes:
* fix: updates Tag model to cascade delete if the Taxonomy or parent Tag is deleted.
* style: adds missing type annotations to rules and python API

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/permissions.py:
```diff
@@ -0,0 +1,27 @@
+        "PATCH": ["%(app_label)s.change_%(model_name)s"],
+        "DELETE": ["%(app_label)s.delete_%(model_name)s"],
+    }
+
+    def has_permission(self, request, view):
+        # Workaround to allow 'retrieve' view to pass through the
+        # method GET permission check. Actual object permission check
+        # is done in the get_object() method.
+        if view.action == "retrieve":
+            return bool(request.user and request.user.is_authenticated)
+        else:
+            return super().has_permission(request, view)
+
```
I can see from Jira that you had a question about permissions, but I don't see it on the PR?

Basically, I'm suggesting that we don't need this method at all, and that only `staff` (or `superuser`s) should be allowed to view taxonomies, or do anything else with them.

pomegranited commented on openedx_tagging/core/tagging/rest_api/v1/permissions.py:
```diff
@@ -0,0 +1,27 @@
+        "PATCH": ["%(app_label)s.change_%(model_name)s"],
+        "DELETE": ["%(app_label)s.delete_%(model_name)s"],
+    }
+
+    def has_permission(self, request, view):
+        # Workaround to allow 'retrieve' view to pass through the
+        # method GET permission check. Actual object permission check
+        # is done in the get_object() method.
+        if view.action == "retrieve":
+            return bool(request.user and request.user.is_authenticated)
+        else:
+            return super().has_permission(request, view)
+
```
I can see from Jira that you had a question about permissions, but I don't see it on the PR?

Basically, I'm suggesting that we don't need this method at all, and so the view permissions will use the rules we've established, namely that that [only `staff` (or `superuser`s) should be allowed to view taxonomies](https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/openedx_tagging/core/tagging/rules.py#L16-L22), or do anything else with them.

Normal users cannot even view them, but that's what this method allows, so it should be removed, and any failing tests updated accordingly.

```suggestion
```

rpenido replied:
The rule (the one you linked) allows ANY user to view an `{enabled=true}` taxonomy. That's why I'm confused. If we want to ensure that only staff should be allowed to view taxonomies, then we should rewrite the rule as follows:

```python
@rules.predicate
def can_view_taxonomy(user: User, taxonomy: Taxonomy = None) -> bool:
    """
    Only taxonomy admins can view a taxonomy.
    """
    return is_taxonomy_admin(user)
```

And this test (line 95):

https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/tests/openedx_tagging/core/tagging/test_rules.py#L82-L96

Sorry for asking for more clarification. Am I misunderstanding you?

pomegranited replied:
@rpenido 

> The rule (the one you linked) allows ANY user to view an {enabled=true} taxonomy. That's why I'm confused.

Ah yes, I spoke incorrectly here.. My apologies! Logged-in users *should* be able to list taxonomies, and so they need non-object view perms.

So instead of hacking around this with `has_permission`, let's fix that rule:

```diff
diff --git a/openedx_tagging/core/tagging/rules.py b/openedx_tagging/core/tagging/rules.py
index 052dfdf..4c506a3 100644
--- a/openedx_tagging/core/tagging/rules.py
+++ b/openedx_tagging/core/tagging/rules.py
@@ -19,7 +19,7 @@ def can_view_taxonomy(user: User, taxonomy: Taxonomy = None) -> bool:
     Anyone can view an enabled taxonomy,
     but only taxonomy admins can view a disabled taxonomy.
     """
-    return (taxonomy and taxonomy.enabled) or is_taxonomy_admin(user)
+    return (not taxonomy or taxonomy.enabled) or is_taxonomy_admin(user)


 @rules.predicate
diff --git a/tests/openedx_tagging/core/tagging/test_rules.py b/tests/openedx_tagging/core/tagging/test_rules.py
index 577c594..415ae19 100644
--- a/tests/openedx_tagging/core/tagging/test_rules.py
+++ b/tests/openedx_tagging/core/tagging/test_rules.py
@@ -90,7 +90,7 @@ class TestRulesTagging(TestTagTaxonomyMixin, TestCase):
         assert self.superuser.has_perm("oel_tagging.view_taxonomy", self.taxonomy)
         assert self.staff.has_perm("oel_tagging.view_taxonomy")
         assert self.staff.has_perm("oel_tagging.view_taxonomy", self.taxonomy)
-        assert not self.learner.has_perm("oel_tagging.view_taxonomy")
+        assert self.learner.has_perm("oel_tagging.view_taxonomy")
         assert (
             self.learner.has_perm("oel_tagging.view_taxonomy", self.taxonomy) == enabled
         )
diff --git a/tests/openedx_tagging/core/tagging/test_views.py b/tests/openedx_tagging/core/tagging/test_views.py
index b714015..e282580 100644
--- a/tests/openedx_tagging/core/tagging/test_views.py
+++ b/tests/openedx_tagging/core/tagging/test_views.py
@@ -86,7 +86,7 @@ class TestTaxonomyViewSet(APITestCase):

     @ddt.data(
         (None, status.HTTP_403_FORBIDDEN),
-        ("user", status.HTTP_403_FORBIDDEN),
+        ("user", status.HTTP_200_OK),
         ("staff", status.HTTP_200_OK),
     )
     @ddt.unpack
```

pomegranited commented on projects/urls.py:
```diff
@@ -9,5 +9,6 @@
     path("media_server/", include("openedx_learning.contrib.media_server.urls")),
     path("rest_api/", include("openedx_learning.rest_api.urls")),
+    path("tagging/rest_api/", include("openedx_tagging.core.tagging.urls")),
     path('__debug__/', include('debug_toolbar.urls')),
```
Ach.. from the [failing test](https://github.com/open-craft/openedx-learning/actions/runs/5603509869/jobs/10250233758?pr=4), it looks like we are stumbling on a missing `debug_toolbar` app in testing.

But that app shouldn't be enabled on a production service, which this line is doing. Could you remove this line please? 
```suggestion
```

pomegranited reviewed (changes requested):
@rpenido  Getting closer :) Few more changes and then it'll be perfect.

And good news -- since https://github.com/openedx/openedx-learning/pull/62 is merged, now, you can prepare your PR against main. Feel free to do that when you're happy with what's here, and you've addressed my remaining changes?

pomegranited committed:
fix: removes COMPRESSED row format from openedx_learning core (#66)

AWS Aurora does not support COMPRESSED in its MySQL implementation:

https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Migrating.RDSMySQL.Import.html

rpenido committed:
feat: add taxonomy view/management apis

rpenido commented:
Closing in favor of https://github.com/openedx/openedx-learning/pull/63
